ratis-examples/src/test/java/org/apache/ratis/examples/filestore/FileStoreStreamingBaseTest.java

```diff
@@ -0,0 +1,133 @@
+    extends BaseTest
+    implements MiniRaftCluster.Factory.Get<CLUSTER> {
+  public static final Logger LOG = LoggerFactory.getLogger(FileStoreStreamingBaseTest.class);
+
+  {
+    final RaftProperties p = getProperties();
+    p.setClass(MiniRaftCluster.STATEMACHINE_CLASS_KEY,
+        FileStoreStateMachine.class, StateMachine.class);
+    ConfUtils.setFile(p::setFile, FileStoreCommon.STATEMACHINE_DIR_KEY,
+        new File(getClassTestDir(), "filestore"));
+  }
+
+  static final int NUM_PEERS = 3;
+  static final int MODULUS = 23;
+
+  @Test
+  public void testFileStoreStream() throws Exception {
+    final CLUSTER cluster = newCluster(NUM_PEERS);
+    cluster.start();
+    RaftServerImpl leader = RaftTestUtil.waitForLeader(cluster);
+
+    final RaftGroup raftGroup = cluster.getGroup();
+    final Collection<RaftPeer> peers = raftGroup.getPeers();
+    Assert.assertEquals(NUM_PEERS, peers.size());
+    RaftPeer raftPeer = peers.iterator().next();
+
+    final CheckedSupplier<FileStoreClient, IOException> newClient =
+        () -> new FileStoreClient(cluster.getGroup(), getProperties(), raftPeer);
+    // TODO: configurable buffer size
+    final int bufferSize = 10_000;
+    testSingleFile("foo", SizeInBytes.valueOf("2M"), bufferSize, newClient, leader);
+
+    cluster.shutdown();
+  }
+
+  private void testSingleFile(
+      String path, SizeInBytes fileLength, int bufferSize, CheckedSupplier<FileStoreClient,
+      IOException> newClient, RaftServerImpl leader)
+      throws Exception {
+    LOG.info("runTestSingleFile with path={}, fileLength={}", path, fileLength);
+    final int size = fileLength.getSizeInt();
+    FileStoreClient client = newClient.get();
+    DataStreamOutput dataStreamOutput = client.getStreamOutput(path);
+    final List<CompletableFuture<DataStreamReply>> futures = new ArrayList<>();
+    final List<Integer> sizes = new ArrayList<>();
+
+    for(int offset = 0; offset < size; ) {
+      final int remaining = size - offset;
+      final int length = Math.min(remaining, bufferSize);
+      final boolean close = length == remaining;
+
+      LOG.trace("write {}, offset={}, length={}, close? {}",
+          path, offset, length, close);
+      final ByteBuffer bf = initBuffer(0, length);
+      futures.add(dataStreamOutput.writeAsync(bf, close));
+      sizes.add(length);
+      offset += length;
+    }
+
+    // TODO: handle when any of the writeAsync has failed.
+    // check writeAsync requests
+    for (int i = 0; i < futures.size(); i++) {
+      final DataStreamReply reply = futures.get(i).join();
+      Assert.assertTrue(reply.isSuccess());
+      Assert.assertEquals(sizes.get(i).longValue(), reply.getBytesWritten());
+      Assert.assertEquals(reply.getType(), i == futures.size() - 1 ? DataStreamPacketHeaderProto.Type.STREAM_DATA_SYNC : DataStreamPacketHeaderProto.Type.STREAM_DATA);
+    }
+
+    dataStreamOutput.closeAsync().join();
+  }
+
+  static ByteBuffer initBuffer(int offset, int size) {
```

Comment:
I think this test is mainly for testing performance, if so, we should create file with large size first, for example 1GB, then use MapByteBuffer to read file and writeAsync(MapByteBuffer). allocateDirect and fill the DirectByteBuffer is not the actual scene we use in ozone.  @szetszwo  what do you think ?

Reply:
The unit test could use generated data in memory.  For the CLI, it should use real files on local disk for running benchmarks.